climpred/relative_entropy.py

```diff
@@ -0,0 +1,217 @@
+from collections import OrderedDict
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import xarray as xr
+from eofs.xarray import Eof
+
+from climpred.prediction import _pseudo_ens
+
+
+def _relative_entropy_formula(sigma_b, sigma_x, mu_x, mu_b, neofs):
+    """
+    Computes the relative entropy formula given in Branstator and Teng, (2010).
+
+    References:
+        - Branstator, Grant, and Haiyan Teng. “Two Limits of Initial-Value
+            Decadal Predictability in a CGCM.” Journal of Climate 23, no. 23
+            (August 27, 2010): 6292–6311. https://doi.org/10/bwq92h.
+        - Kleeman, Richard. “Measuring Dynamical Prediction Utility Using
+            Relative Entropy.” Journal of the Atmospheric Sciences 59, no. 13
+            (July 1, 2002): 2057–72. https://doi.org/10/fqwxpk.
+
+    Args:
+        sigma_b (xr.DataArray): covariance matrix of baseline distribution
+        sigma_x (xr.DataArray): covariance matrix of forecast distribution
+        mu_b (xr.DataArray): mean state vector of the baseline distribution
+        mu_x (xr.DataArray): mean state vector of the forecast distribution
+        neofs (int): number of EOFs to use
+
+    Returns:
+        R (float): relative entropy
+        dispersion (float): dispersion component
+        signal (float): signal component
+    """
+    fac = 0.5
+    dispersion = fac * (np.log(np.linalg.det(sigma_b) / np.linalg.det(sigma_x))
+                        + np.trace(sigma_x / sigma_b) - neofs)
+
+    # https://stackoverflow.com/questions/7160162/left-matrix-division-and-numpy-solve
+    # (A.T\B)*A
+    x, resid, rank, s = np.linalg.lstsq(
+        sigma_b, mu_x - mu_b)  # sigma_b \ (mu_x - mu_b)
+    signal = fac * np.matmul((mu_x - mu_b), x)
+    R = dispersion + signal
+    return R, dispersion, signal
+
+
+def _gen_control_ensemble(ds, control, it=10):
```

Comment:
I want to bootstrap it times the same calculation as I do with init ensemble with an uninit ensemble and then take the sig th percentile

Reply:
ah sorry. here in this context I generate a very large ensemble with say 100 members. therefore I can construct my control convariance matrix as 100x100. if I just do the 10x10 uninit control matrix the results differ from time to time and I get very large spikes in my results.

Reply:
`it` is basically the number of times I get the uninitialized ensemble into an `it*ds.member.size`-member uninit ensemble. not sure how much this implementation is like the people thought of it in the first place

Reply:
I dont have a LENS, therefore this way I create a LENS it-times larger than ds.

Reply:
my top-level function isn't designed for dple, lens now, as lens has no initializations.

LENS dims: member 1-40, time 1950-2015
DPLE dims: member 1-40, init 1950-2015, time: 1-10

Solutions: 
a) I create a new control distribution and adapt my compute_rel_ent
b) we create a separate compute_rel_ent_hindcast

Changing or a new compute_hindcast function would involve a different for loop for uninitialized_control or some rename of dimensions before.

now for uninitialized with init dimension (sounds contradictory):
```python
    for init in initializations: #inits
        for t in initialized.time.values[:ntime]: #lead times
```
or maybe we write a function that creates an LENS function with a randomized init dim

then with original LENS input of dims member (which we cannot touch is we need it for the PCs) and time:
```python
    for time in uninitialized.time: # different start dates 
        for ly in lead_years:
            t = t + ly
```
I hope you get the problem I face here. I am not sure whether the second code snippet would be a solution for a DPLE, LENS implementation.

I don't want to implement this before you comment on this, because it seems technically annoying.